### ISSUE TITLE
Why are there only 15 Macro

### ISSUE THREAD

**Issue by Benjamin-Blackswan** [the-via/keyboards#632]
firstly thank you to all those involved with VIA and bring programmability to those of us too scared to use qmk.

I ordered nk65 milkshake 3 month again, after reading about this new thing call VIA that gives the power of programmability without having to deal with qmk.

I just got my nk65 yesterday and used VIA for the first time and I have to say I am really disappointed with the programmability and functionality of VIA to say the least. It only 4 layers in total and only 15 macro?

I paid $300 AUD for the nk65 as a replacement over my po3ker. Yet it is weaker than my 5 year old board with no real programmability.

There is nothing on your website caniusevia.com that says that VIA has only 4 layer and 15 macro. If you have mentioned that VIA has only 4 layer and 15 macro I would never have bought the nk65. It offers no additional benefit over my pok3r.

VIA should NOT be promoted as a qmk alternative by any KB manufacturers if it has virtually no programmability.

I know VIA has no control over KB manufacturers but on caniusevia.com you should make it clear of the limitations so consumers dont wasted money and time only to find that VIA is no where close to the power of qmk.

**Comment by Nietzche2021** [comment]
that was by me. I had to delete that account for personally reasons

**Comment by itspngu** [comment]
> that was by me. I had to delete that account for personally reasons

Why do you need more than 4 layers, hell, the Fuhrer pretty much only had one usable hand and got along.

**Comment by yiancar** [comment]
1. VIA is a QMK companion. It is **not** a QMK replacement.
2. The keyboard you have mentioned and all the keyboards supported by VIA run QMK. You can gladly program QMK manually and get up to 32 layers. Similarly you can get as many macros as you want.
3. All your other statements are confusing and inaccurate.
4. VIA is designed for entry level use of QMK. A user who wants to use more advance features of QMK can freely recompile the firmware with any additions they would like.
5. I personally encourage you to check your facts before making inaccurate statements. Both QMK and VIA are developed by a community of members without any monetary compensation. Blaming culture has never got anyone anywhere.

I dont think this issue can be explored further especially as the original creator has deleted their account.

**Comment by Nietzche2021** [comment]
> the Fuhrer pretty much only had one usable hand and got along.

this is a blood version of the Reichskriegsflagge of the Second Reich, its got nothing to do with the Nazi or the far right, please educate yourself.

**Comment by yiancar** [comment]
Beautiful edit. Thank you for the drama

**Comment by Nietzche2021** [comment]
mate I still dont understand why there are only FTFTEEN (15) macros?  That's not enough for even one app

**Comment by itspngu** [comment]
> well the Reichskriegsflagge of the second German Empire has nothing to do with Hitler

Wilhelm II's left arm was crippled at birth. He, too, got along. You seem to have a valid use case here and I'd like to see this issue being looked into further. Godspeed!

**Comment by Nietzche2021** [comment]
> Wilhelm II's left arm was crippled at birth. He, too, got along.

brah, he too lost a world war for Germany. If he had unlimited macro and more layers, then he might have won WW1 and then there would be no Hitler

**Comment by drashna** [comment]
Both are a limitation on the amount of EEPROM available to the microcontroller (3-4 layers well use up most of the EEPROM, actually).  As for stuff like Control+Alt+Del for instance can be done *in* the keymap without the need of a macro.  Eg, If you use the ANY key, you can add C+A+Del by using `C(A(KC_DEL))`

Tone matters.  Ranting about an issue rather than asking isn't a good way to get help.

**Comment by Nietzche2021** [comment]
> Both are a limitation on the amount of EEPROM available to the microcontroller (3-4 layers well use up most of the EEPROM, actually). 

So EEPROM limitation affects only VIA but not QMK?

What you're saying is there is no hope of VIA getting any more layers and macro and I will have to head down the QMK route if I want more layer/macro?

**Comment by yiancar** [comment]
As @drashna was kind enough to answer I will continue his explanation.

Lets forget VIA for a second.
When a layer is set in the qmk firmware and compiled it is fixed. What I mean by this is that that layer is saved in the flash memory of the chip (think of this as a cd, all cd players can read the contents but here is a special procedure (compilation and flashing in our case) to write it).

QMK offers you the capability of **dynamically** setting layers as well. This means, that the layers can be changed without recompilation or flashing, IE on the fly. To do this the layers should be stored somewhere easy to read and write. in our case this is in the eeprom (more like a hardisk where it is easy to read and write).

As @drashna explained above, due to the limited size of the eeprom, one can only store 4 layers which can be dynamically changed.

Now, VIA, is an interface tool, which tells QMK what to set this dynamic layers to.

TL;DR When compiling QMK one can use up to 32 layers and "infinite" macros, those layers and macros are fixed and cannot be changed dynamically, on the fly. Alternatively one can use 4 dynamic layers and 15 dynamic macros and change them on the fly using VIA.

**Comment by Nietzche2021** [comment]
thank you yiancar, that was a good ELI5 explanation. 
Due to the hardware limitation, I guess I will have too go down the qmk route.

**Comment by Shaglock** [comment]
thank you @yiancar for explanation, I was trying to figure out why my 5th layer worked on QMK and didn't work with VIA. I guess that's the reason :)
